src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DataAnnotationsClientModelValidatorProvider.cs

```diff
@@ -0,0 +1,105 @@
+        internal Dictionary<Type, DataAnnotationsClientModelValidationFactory> AttributeFactories
+        {
+            get { return _attributeFactories; }
+        }
+
+        /// <inheritdoc />
+        public void GetValidators(ClientValidatorProviderContext context)
+        {
+            foreach (var attribute in context.ValidatorMetadata.OfType<ValidationAttribute>())
+            {
+                DataAnnotationsClientModelValidationFactory factory;
+                if (!_attributeFactories.TryGetValue(attribute.GetType(), out factory))
+                {
+                    factory = _defaultAttributeFactory;
+                }
+
+                context.Validators.Add(factory(attribute));
+            }
+        }
+
+        private static Dictionary<Type, DataAnnotationsClientModelValidationFactory> BuildAttributeFactoriesDictionary()
+        {
+            var dict = new Dictionary<Type, DataAnnotationsClientModelValidationFactory>();
+            AddValidationAttributeAdapter(dict, typeof(RegularExpressionAttribute),
+                (attribute) => new RegularExpressionAttributeAdapter((RegularExpressionAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(MaxLengthAttribute),
+                (attribute) => new MaxLengthAttributeAdapter((MaxLengthAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(MinLengthAttribute),
+                (attribute) => new MinLengthAttributeAdapter((MinLengthAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(CompareAttribute),
+                (attribute) => new CompareAttributeAdapter((CompareAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(RequiredAttribute),
+                (attribute) => new RequiredAttributeAdapter((RequiredAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(RangeAttribute),
+                (attribute) => new RangeAttributeAdapter((RangeAttribute)attribute));
+
+            AddValidationAttributeAdapter(dict, typeof(StringLengthAttribute),
+                (attribute) => new StringLengthAttributeAdapter((StringLengthAttribute)attribute));
+
+            AddDataTypeAttributeAdapter(dict, typeof(CreditCardAttribute), "creditcard");
+            AddDataTypeAttributeAdapter(dict, typeof(EmailAddressAttribute), "email");
+            AddDataTypeAttributeAdapter(dict, typeof(PhoneAttribute), "phone");
+            AddDataTypeAttributeAdapter(dict, typeof(UrlAttribute), "url");
+
+            return dict;
+        }
+
+        private static void AddValidationAttributeAdapter(
+            Dictionary<Type, DataAnnotationsClientModelValidationFactory> dictionary,
+            Type validationAttributeType,
+            DataAnnotationsClientModelValidationFactory factory)
+        {
+            if (validationAttributeType != null)
+            {
+                dictionary.Add(validationAttributeType, factory);
+            }
+        }
```

Comment:
Would this be simpler to write as a dictionary initializer? Those didn't exist when this code was originally written. Your call if you think it's better or not.